src/components/BarChart/index.tsx

```diff
@@ -0,0 +1,102 @@
+import React, { useRef, useEffect } from 'react';
+import { _cs } from '@togglecorp/fujs';
+
+import { select } from 'd3-selection';
+import { scaleBand, scaleLinear } from 'd3-scale';
+import { axisBottom, axisLeft } from 'd3-axis';
+import { max } from 'd3-array';
+
+import styles from './styles.css';
+
+interface Margin {
+    top: number;
+    right: number;
+    bottom: number;
+    left: number;
+}
+interface Props<K> {
+    width: number;
+    height: number;
+    data: K[];
+    valueSelector: (v: K) => number;
+    labelSelector: (v: K) => string;
+    colorSelector: (v: K) => string;
+    margin: Margin;
+}
+
+const defaultProps = {
+    colorSelector: () => '#b4d9cc',
+    margin: { top: 30, right: 0, bottom: 40, left: 40 },
+};
+
+function BarChart<K>(props: Props<K>) {
+    const {
+        width,
+        height,
+        data,
+        valueSelector,
+        labelSelector,
+        colorSelector,
+        margin: {
+            top,
+            right,
+            bottom,
+            left,
+        },
+    }  = props;
+
+    const svgRef = useRef(null);
+
+    useEffect(() => {
+        const svg = select(svgRef.current)
+            .attr('width', width)
+            .attr('height', height);
+
+        const x = scaleBand()
+            .domain(data.map(d => labelSelector(d)))
+            .range([left, width - right])
+            .padding(0.1);
+
+        const y = scaleLinear()
+            .domain([0, max(data, valueSelector)])
+            .nice()
+            .range([height - bottom, top]);
+
+        const xAxis = axisBottom(x);
+
+        const yAxis = axisLeft(y);
```

Comment:
can we not just memoize these?

Reply:
i don't know if it's better to use separate useEffect hooks for different side-effects (d3 invocation)